engines/director/lingo/lingo-bytecode.cpp

```diff
@@ -0,0 +1,311 @@
+	{ 0x06, Lingo::c_sub, "" },
+	{ 0x07, Lingo::c_div, "" },
+	{ 0x08, Lingo::c_mod, "" },
+	{ 0x09, Lingo::c_negate, "" },
+	{ 0x0a, Lingo::c_ampersand, "" },
+	{ 0x0b, Lingo::c_concat, "" },
+	{ 0x0c, Lingo::c_lt, "" },
+	{ 0x0d, Lingo::c_le, "" },
+	{ 0x0e, Lingo::c_neq, "" },
+	{ 0x0f, Lingo::c_eq, "" },
+	{ 0x10, Lingo::c_gt, "" },
+	{ 0x11, Lingo::c_ge, "" },
+	{ 0x12, Lingo::c_and, "" },
+	{ 0x13, Lingo::c_or, "" },
+	{ 0x14, Lingo::c_not, "" },
+	{ 0x15, Lingo::c_contains, "" },
+	{ 0x16, Lingo::c_starts, "" },
+	{ 0x17, Lingo::c_of, "" },
+	{ 0x18, Lingo::c_hilite, "" },
+	{ 0x19, Lingo::c_intersects, "" },
+	{ 0x1a, Lingo::c_within, "" },
+	{ 0x1b, Lingo::c_field, "" },
+	{ 0x1c, Lingo::c_tell, "" },
+	{ 0x1d, Lingo::c_telldone, "" },
+	{ 0x41, Lingo::c_intpush, "b" },
+	{ 0, 0, 0 }
+};
+
+void Lingo::initBytecode() {
+	for (LingoV4Bytecode *op = lingoV4; op->opcode; op++) {
+		_lingoV4[op->opcode] = new Opcode( op->func, op->proto );
+	}
+}
+
+
+void Lingo::addCodeV4(Common::SeekableSubReadStreamEndian &stream, ScriptType type, uint16 id) {
+	debugC(1, kDebugLingoCompile, "Add V4 bytecode for type %s with id %d", scriptType2str(type), id);
+
+	if (_scriptContexts[type].contains(id)) {
+		for (size_t j = 0; j < _scriptContexts[type][id]->functions.size(); j++) {
+			delete _scriptContexts[type][id]->functions[j];
+		}
+		delete _scriptContexts[type][id];
+	}
+
+	_currentScriptContext = new ScriptContext;
+	_currentScriptType = type;
+	_currentEntityId = id;
+	_scriptContexts[type][id] = _currentScriptContext;
+
+	// read the Lscr header!
+	// unk1
+	for (uint32 i = 0; i < 0x10; i++) {
+		stream.readByte();
+	}
+	uint16 codeStoreOffset = stream.readUint16();
+	// unk2
+	for (uint32 i = 0; i < 0x2e; i++) {
+		stream.readByte();
+	}
+	uint16 globalsOffset = stream.readUint16();
+	uint16 globalsCount = stream.readUint16();
+	// unk3
+	for (uint32 i = 0; i < 0x4; i++) {
+		stream.readByte();
+	}
+	uint16 functionsCount = stream.readUint16();
+	stream.readUint16();
+	uint16 functionsOffset = stream.readUint16();
+	uint16 constsCount = stream.readUint16();
+	stream.readUint16();
+	uint16 constsOffset = stream.readUint16();
+	stream.readUint16();
+	stream.readUint16();
+	stream.readUint16();
+	uint16 constsBase = stream.readUint16();
+
+	// preload all the constants!
+	// these are stored as a reference table of 6 byte entries, followed by a storage area.
+
+	// copy the storage area first.
+	uint32 constsStoreOffset = constsOffset + 6 * constsCount;
+	uint32 constsStoreSize = stream.size() - constsStoreOffset;
+	stream.seek(constsStoreOffset);
+	byte *constsStore = (byte *)malloc(constsStoreSize);
+	stream.read(constsStore, constsStoreSize);
+
+	Common::Array<Datum> constsData;
+
+	// read each entry in the reference table.
+	stream.seek(constsOffset);
+	for (uint16 i=0; i < constsCount; i++) {
+		Datum constant;
+		uint16 constType = stream.readUint16();
+		uint32 value = stream.readUint32();
+		switch (constType) {
+			case 1: // String type
```

Comment:
Indentation of `case` statements must be aligned with their `switch` counterparts as per ou code formatting conventions.

Reply:
Fixed